File: src/searchat/config/path_resolver.py
"""
Cross-platform path resolution for AI coding agent conversation directories.

Supports:
- Claude Code: ~/.claude/projects/**/*.jsonl
- Codex: ~/.codex/sessions/**/*.jsonl
- Mistral Vibe: ~/.vibe/logs/session/*.json

Handles path translation between:
- Windows native paths (C:/Users/...)
- WSL paths (/mnt/c/Users/... or //wsl$/Ubuntu/...)
- Unix/Linux/macOS paths (/home/user/...)
"""

import os
import sys
import platform
from pathlib import Path
from typing import List, Optional

from .constants import (
    ENV_DATA_DIR,
    ENV_WINDOWS_PROJECTS,
    ENV_WSL_PROJECTS,
    ENV_ADDITIONAL_DIRS,
    CLAUDE_DIR_CANDIDATES,
    CODEX_DIR_CANDIDATES,
)


class PathResolver:
    """Resolves and translates paths across different platforms."""

    @staticmethod
    def safe_exists(path: Path) -> bool:
        """Return False instead of raising for inaccessible paths."""
        try:
            return path.exists()
        except OSError:
            return False

    @staticmethod
    def safe_resolve(path: Path) -> Path:
        """Resolve a path when possible without failing on broken UNC shares."""
        try:
            return path.resolve() if PathResolver.safe_exists(path) else path
        except OSError:
            return path

    @staticmethod
    def expand_path_template(path: str) -> str:
        """
        Expand path templates with environment variables.

        Supports:
        - {username} -> current username
        - {home} -> user home directory
        - Environment variables: $VAR or ${VAR}

        Args:
            path: Path template string

        Returns:
            Expanded path string
        """
        # Expand custom placeholders
        path = path.replace("{username}", os.getenv("USERNAME") or os.getenv("USER") or "unknown")
        path = path.replace("{home}", str(Path.home()))

        # Expand ~ to home directory
        path = os.path.expanduser(path)

        # Expand environment variables
        path = os.path.expandvars(path)

        return path

    @staticmethod
    def get_shared_search_dir(config=None) -> Path:
        """
        Get the shared search directory path.

        Checks in order:
        1. Environment variable (SEARCHAT_DATA_DIR)
        2. Config file setting
        3. Default location

        Args:
            config: Optional Config object

        Returns:
            Path to search directory
        """
        # Check environment variable first
        env_dir = os.getenv(ENV_DATA_DIR)
        if env_dir:
            return Path(PathResolver.expand_path_template(env_dir))

        # Load config if not provided
        if config is None:
            from searchat.config import Config
            config = Config.load()

        # Expand template in config path
        search_dir = PathResolver.expand_path_template(config.paths.search_directory)
        return Path(search_dir)

    @staticmethod
    def detect_platform() -> str:
        """
        Detect the current platform.

        Returns:
            One of: "windows", "wsl", "linux", "macos"
        """
        if sys.platform == "win32":
            return "windows"
        elif sys.platform == "darwin":
            return "macos"
        elif sys.platform.startswith("linux"):
            # Check if we're in WSL
            if PathResolver._is_wsl():
                return "wsl"
            return "linux"
        return "unknown"

    @staticmethod
    def _is_wsl() -> bool:
        """Check if running under WSL."""
        try:
            with open("/proc/version", "r") as f:
                return "microsoft" in f.read().lower()
        except (FileNotFoundError, PermissionError):
            return False

    @staticmethod
    def translate_claude_path(path: str) -> Path:
        """
        Translate path between different platform conventions.

        Handles:
        - Windows <-> WSL path translation
        - Network UNC paths (//wsl$/...)
        - Unix mount points (/mnt/c/...)

        Args:
            path: Path string to translate

        Returns:
            Translated Path object
        """
        current_platform = PathResolver.detect_platform()

        # Windows native -> WSL mount point
        if current_platform in ("wsl", "linux") and (path[1:3] == ":\\" or path[1:3] == ":/"):
            # C:\Users\... -> /mnt/c/Users/...
            drive = path[0].lower()
            rest = path[3:].replace("\\", "/")
            return Path(f"/mnt/{drive}/{rest}")

        # WSL mount point -> Windows native
        elif current_platform == "windows" and path.startswith("/mnt/"):
            # /mnt/c/Users/... -> C:\Users\...
            parts = path[5:].split("/", 1)
            if len(parts) == 2:
                drive, rest = parts
                return Path(f"{drive.upper()}:\\{rest.replace('/', chr(92))}")
            else:
                drive = parts[0]
                return Path(f"{drive.upper()}:\\")

        # WSL UNC path (\\wsl$\Ubuntu\home\...) - already in Windows format
        elif current_platform == "windows" and path.startswith("\\\\wsl"):
            return Path(path)

        # No translation needed
        return Path(path)

    @staticmethod
    def resolve_claude_dirs(config=None) -> List[Path]:
        """
        Resolve all accessible Claude conversation directories.

        Checks:
        1. Additional directories from environment variable
        2. Configured Windows directory
        3. Configured WSL directory
        4. Standard fallback locations

        Args:
            config: Optional Config object

        Returns:
            List of accessible Path objects

        Raises:
            RuntimeError: If no directories are found
        """
        if config is None:
            from searchat.config import Config
            config = Config.load()

        paths = []
        current_platform = PathResolver.detect_platform()

        # 1. Check additional directories from environment
        additional_dirs = os.getenv(ENV_ADDITIONAL_DIRS)
        if additional_dirs:
            for dir_path in additional_dirs.split(os.pathsep):
                expanded = PathResolver.expand_path_template(dir_path)
                path = PathResolver.translate_claude_path(expanded)
                if PathResolver.safe_exists(path):
                    paths.append(path)

        # 2. Check Windows directory
        windows_dir = PathResolver.expand_path_template(config.paths.claude_directory_windows)
        windows_path = PathResolver.translate_claude_path(windows_dir)
        if PathResolver.safe_exists(windows_path):
            paths.append(windows_path)

        # 3. Check WSL directory (only from Windows)
        if current_platform == "windows":
            wsl_dir = PathResolver.expand_path_template(config.paths.claude_directory_wsl)
            if wsl_dir.strip():
                wsl_path = Path(wsl_dir)
                # Note: Path.exists() doesn't work reliably with WSL network paths,
                # but glob/iterdir work fine, so always add the path on Windows
                paths.append(wsl_path)

        # 4. Check standard fallback locations
        if not paths:
            for candidate in CLAUDE_DIR_CANDIDATES:
                if PathResolver.safe_exists(candidate):
                    paths.append(candidate)

        # Remove duplicates while preserving order
        seen = set()
        unique_paths = []
        for path in paths:
            # Resolve to absolute path for comparison
            resolved = PathResolver.safe_resolve(path)
            path_str = str(resolved)
            if path_str not in seen:
                seen.add(path_str)
                unique_paths.append(path)

        if not unique_paths:
            searched = [
                PathResolver.expand_path_template(config.paths.claude_directory_windows),
                PathResolver.expand_path_template(config.paths.claude_directory_wsl),
            ] + [str(c) for c in CLAUDE_DIR_CANDIDATES]

            raise RuntimeError(
                f"No Claude conversation directories found.\n"
                f"Platform: {current_platform}\n"
                f"Searched locations:\n" +
                "\n".join(f"  - {p}" for p in searched)
            )

        return unique_paths

    @staticmethod
    def ensure_directory(path: Path) -> Path:
        """
        Ensure a directory exists, creating it if necessary.

        Args:
            path: Directory path to ensure

        Returns:
            The path object

        Raises:
            PermissionError: If directory cannot be created
        """
        path.mkdir(parents=True, exist_ok=True)
        return path

    @staticmethod
    def resolve_vibe_dirs() -> List[Path]:
        """
        Resolve Mistral Vibe session directories.

        Vibe stores sessions at ~/.vibe/logs/session/

        Returns:
            List of accessible Vibe session directories
        """
        paths = []

        # Standard Vibe location
        vibe_session_dir = Path.home() / ".vibe" / "logs" / "session"
        if PathResolver.safe_exists(vibe_session_dir):
            paths.append(vibe_session_dir)

        # Check VIBE_HOME environment variable
        vibe_home = os.getenv("VIBE_HOME")
        if vibe_home:
            custom_session_dir = Path(vibe_home) / "logs" / "session"
            if PathResolver.safe_exists(custom_session_dir) and custom_session_dir not in paths:
                paths.append(custom_session_dir)

        return paths

    @staticmethod
    def resolve_codex_dirs() -> List[Path]:
        """
        Resolve Codex session directories.

        Codex stores sessions at ~/.codex/sessions/YYYY/MM/DD/*.jsonl

        Returns:
            List of accessible Codex session directories
        """
        paths = []
        current_platform = PathResolver.detect_platform()

        for candidate in CODEX_DIR_CANDIDATES:
            if PathResolver.safe_exists(candidate):
                paths.append(candidate)

        if current_platform == "windows":
            username = os.getenv("USERNAME") or os.getenv("USER") or "user"
            wsl_candidates = [
                Path(f"\\\\wsl.localhost\\Ubuntu\\home\\{username}\\.codex\\sessions"),
                Path(f"\\\\wsl$\\Ubuntu\\home\\{username}\\.codex\\sessions"),
            ]
            for candidate in wsl_candidates:
                if candidate not in paths:
                    paths.append(candidate)

        seen = set()
        unique_paths = []
        for path in paths:
            resolved = PathResolver.safe_resolve(path)
            path_str = str(resolved)
            if path_str not in seen:
                seen.add(path_str)
                unique_paths.append(path)

        return unique_paths

    @staticmethod
    def resolve_all_agent_dirs(config=None) -> dict:
        """
        Resolve all supported AI coding agent directories.

        Returns:
            Dict mapping agent name to list of directories:
            {
                'claude': [Path(...), ...],
                'codex': [Path(...), ...],
                'vibe': [Path(...), ...]
            }
        """
        return {
            'claude': PathResolver.resolve_claude_dirs(config),
            'codex': PathResolver.resolve_codex_dirs(),
            'vibe': PathResolver.resolve_vibe_dirs(),
        }

File: rust/crates/searchat-config/src/path_resolver.rs
use std::path::{Path, PathBuf};

use crate::constants::*;

// ============================================================================
// Platform detection
// ============================================================================

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Platform {
    Windows,
    Wsl,
    Linux,
    Macos,
    Unknown,
}

/// Detect the current runtime platform.
pub fn detect_platform() -> Platform {
    if cfg!(target_os = "windows") {
        Platform::Windows
    } else if cfg!(target_os = "macos") {
        Platform::Macos
    } else if cfg!(target_os = "linux") {
        if is_wsl() {
            Platform::Wsl
        } else {
            Platform::Linux
        }
    } else {
        Platform::Unknown
    }
}

/// Check whether the process is running under WSL by inspecting /proc/version.
pub fn is_wsl() -> bool {
    match std::fs::read_to_string("/proc/version") {
        Ok(content) => {
            let lower = content.to_lowercase();
            lower.contains("microsoft") || lower.contains("wsl")
        }
        Err(_) => false,
    }
}

// ============================================================================
// Standard directory helpers
// ============================================================================

/// `~/.searchat/`
pub fn default_data_dir() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join(CONFIG_DIR_NAME)
}

/// `~/.searchat/config/`
pub fn default_config_dir() -> PathBuf {
    default_data_dir().join(DEFAULT_CONFIG_SUBDIR)
}

/// `~/.searchat/data/`
pub fn default_storage_dir() -> PathBuf {
    default_data_dir().join(DEFAULT_DATA_SUBDIR)
}

/// `~/.searchat/logs/`
pub fn default_logs_dir() -> PathBuf {
    default_data_dir().join(DEFAULT_LOGS_SUBDIR)
}

/// `~/.claude/projects/`
pub fn default_claude_projects_dir() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join(CLAUDE_DIR_NAME)
        .join(CLAUDE_PROJECTS_SUBDIR)
}

/// `~/.codex/sessions/`
pub fn default_codex_sessions_dir() -> PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join(CODEX_DIR_NAME)
        .join(CODEX_SESSIONS_SUBDIR)
}

// ============================================================================
// Path translation (Windows <-> WSL)
// ============================================================================

/// Translate a path string between Windows and WSL conventions, based on the
/// current platform.
///
/// - On WSL/Linux: `C:\Users\...` or `C:/Users/...` → `/mnt/c/Users/...`
/// - On Windows:   `/mnt/c/Users/...` → `C:\Users\...`
/// - Otherwise:    return as-is
pub fn translate_path(path: &str) -> PathBuf {
    let platform = detect_platform();

    // Windows drive path seen from WSL
    if matches!(platform, Platform::Wsl | Platform::Linux) {
        if path.len() >= 3 {
            let bytes = path.as_bytes();
            let second = bytes.get(1).copied();
            let third = bytes.get(2).copied();
            if second == Some(b':') && (third == Some(b'\\') || third == Some(b'/')) {
                let drive = path.chars().next().unwrap().to_lowercase().next().unwrap();
                let rest = path[3..].replace('\\', "/");
                return PathBuf::from(format!("/mnt/{}/{}", drive, rest));
            }
        }
    }

    // WSL mount point seen from Windows
    if platform == Platform::Windows && path.starts_with("/mnt/") {
        let stripped = &path[5..]; // remove "/mnt/"
        let mut parts = stripped.splitn(2, '/');
        let drive = parts.next().unwrap_or("c");
        let rest = parts.next().unwrap_or("");
        return PathBuf::from(format!("{}:\\{}", drive.to_uppercase(), rest.replace('/', "\\")));
    }

    // WSL UNC path (\\wsl$\...) — already Windows-formatted
    // No translation needed; pass through.
    PathBuf::from(path)
}

// ============================================================================
// Safe filesystem helpers
// ============================================================================

/// Returns `false` instead of panicking for inaccessible paths (e.g. UNC shares).
pub fn safe_exists(path: &Path) -> bool {
    path.try_exists().unwrap_or(false)
}

/// Resolve a path to absolute without panicking on inaccessible network shares.
pub fn safe_resolve(path: &Path) -> PathBuf {
    if safe_exists(path) {
        path.canonicalize().unwrap_or_else(|_| path.to_path_buf())
    } else {
        path.to_path_buf()
    }
}

/// Ensure a directory exists, creating it (and parents) if needed.
pub fn ensure_directory(path: &Path) -> std::io::Result<()> {
    std::fs::create_dir_all(path)
}

// ============================================================================
// Agent directory resolution
// ============================================================================

/// Resolve all accessible Claude conversation directories in priority order:
/// 1. `SEARCHAT_ADDITIONAL_DIRS` (colon/semicolon-separated on the current OS)
/// 2. Windows project dir from config (translated if needed)
/// 3. WSL project dir from config (Windows only)
/// 4. Standard fallback candidates
pub fn resolve_claude_dirs(config: &crate::settings::Config) -> Vec<PathBuf> {
    let mut paths: Vec<PathBuf> = Vec::new();
    let platform = detect_platform();

    // 1. Additional dirs from environment
    if let Ok(additional) = std::env::var(ENV_ADDITIONAL_DIRS) {
        let sep = if platform == Platform::Windows { ';' } else { ':' };
        for dir in additional.split(sep) {
            let expanded = expand_path_template(dir);
            let translated = translate_path(&expanded);
            if safe_exists(&translated) {
                paths.push(translated);
            }
        }
    }

    // 2. Configured Windows directory
    let windows_expanded = expand_path_template(&config.paths.claude_directory_windows);
    let windows_path = translate_path(&windows_expanded);
    if safe_exists(&windows_path) {
        paths.push(windows_path);
    }

    // 3. WSL directory (only relevant from Windows)
    if platform == Platform::Windows {
        let wsl = config.paths.claude_directory_wsl.trim();
        if !wsl.is_empty() {
            // UNC-style WSL paths don't respond to Path::exists reliably on Windows,
            // so always include them as candidates.
            paths.push(PathBuf::from(wsl));
        }
    }

    // 4. Standard fallback
    if paths.is_empty() {
        let fallback = default_claude_projects_dir();
        if safe_exists(&fallback) {
            paths.push(fallback);
        }
    }

    deduplicate_paths(paths)
}

/// Resolve Codex session directories.
pub fn resolve_codex_dirs() -> Vec<PathBuf> {
    let mut paths: Vec<PathBuf> = Vec::new();
    let platform = detect_platform();

    let standard = default_codex_sessions_dir();
    if safe_exists(&standard) {
        paths.push(standard);
    }

    if platform == Platform::Windows {
        let username = std::env::var("USERNAME")
            .or_else(|_| std::env::var("USER"))
            .unwrap_or_else(|_| "user".to_string());
        let wsl_candidates = vec![
            PathBuf::from(format!(
                "\\\\wsl.localhost\\Ubuntu\\home\\{}\\{}\\{}",
                username, CODEX_DIR_NAME, CODEX_SESSIONS_SUBDIR
            )),
            PathBuf::from(format!(
                "\\\\wsl$\\Ubuntu\\home\\{}\\{}\\{}",
                username, CODEX_DIR_NAME, CODEX_SESSIONS_SUBDIR
            )),
        ];
        for candidate in wsl_candidates {
            paths.push(candidate);
        }
    }

    deduplicate_paths(paths)
}

/// Resolve Vibe session directories (`~/.vibe/logs/session/`).
pub fn resolve_vibe_dirs() -> Vec<PathBuf> {
    let mut paths: Vec<PathBuf> = Vec::new();

    let standard = dirs::home_dir()
        .unwrap_or_default()
        .join(".vibe")
        .join("logs")
        .join("session");
    if safe_exists(&standard) {
        paths.push(standard);
    }

    if let Ok(vibe_home) = std::env::var("VIBE_HOME") {
        let custom = PathBuf::from(&vibe_home).join("logs").join("session");
        if safe_exists(&custom) && !paths.contains(&custom) {
            paths.push(custom);
        }
    }

    paths
}

// ============================================================================
// Helpers
// ============================================================================

/// Expand `{username}`, `{home}`, `~`, and environment variables in a path string.
pub fn expand_path_template(path: &str) -> String {
    let username = std::env::var("USERNAME")
        .or_else(|_| std::env::var("USER"))
        .unwrap_or_else(|_| "unknown".to_string());
    let home = dirs::home_dir()
        .unwrap_or_default()
        .to_string_lossy()
        .into_owned();

    let mut result = path.to_string();
    result = result.replace("{username}", &username);
    result = result.replace("{home}", &home);
    result = result.replace('~', &home);
    result
}

/// Deduplicate a list of paths while preserving order, using canonical form for comparison.
fn deduplicate_paths(paths: Vec<PathBuf>) -> Vec<PathBuf> {
    let mut seen = std::collections::HashSet::new();
    let mut unique = Vec::new();
    for path in paths {
        let key = safe_resolve(&path).to_string_lossy().into_owned();
        if seen.insert(key) {
            unique.push(path);
        }
    }
    unique
}
